### PR TITLE
GH-85168: Use filesystem encoding when converting to/from `file` URIs

### PR DESCRIPTION
**Pull request by barneygale**
Adjust `urllib.request.url2pathname()` and `pathname2url()` to use the filesystem encoding when quoting and unquoting file URIs, rather than forcing use of UTF-8.

No changes are needed in the `nturl2path` module because Windows always uses UTF-8, per PEP 529.

* Issue: gh-85168

### CODE REVIEW

**Comment by barneygale** [comment]
The [freedesktop file URI spec](https://edeproject.org/spec/file-uri-spec.txt) is a useful reference here:

> There is no specified encoding for filenames, and although we hope
that eventually all filenames will be encoded in UTF8 we can't rely on
this, because then we would be unable to e.g. rename a file with a
misencoded filename.

Which isn't completely explicit, but the example at the end implies the local filesystem encoding is intended.

File: Lib/urllib/request.py [hunk 1]
```diff
@@ -1660,12 +1660,12 @@ def url2pathname(pathname):
             # URL has an empty authority section, so the path begins on the
             # third character.
             pathname = pathname[2:]
-        return unquote(pathname)
+        return os.fsdecode(unquote_to_bytes(pathname))
```
Comment (serhiy-storchaka) [inline Lib/urllib/request.py]: `unquote_to_bytes()` encodes non-ASCII characters as UTF-8. Then `os.fsdecode()` decodes it using filesystem encoding.

You can specify `encoding` and `errors` for `unquote()`.

File: Lib/test/test_urllib.py [hunk 2]
```diff
@@ -1561,6 +1557,13 @@ def test_pathname2url_posix(self):
         self.assertEqual(fn('/'), '/')
         self.assertEqual(fn('/a/b.c'), '/a/b.c')
         self.assertEqual(fn('/a/b%#c'), '/a/b%25%23c')
+        try:
+            expect = os.fsencode('\xe9')
```
Comment (serhiy-storchaka) [inline Lib/test/test_urllib.py]: You can use `os_helper.FS_NONASCII`.

**Review by serhiy-storchaka — COMMENTED** [review]
LGTM in general. But we should support URIs containing not quoted non-ASCII characters.

**Review by serhiy-storchaka — APPROVED** [review]
LGTM.

You can also use `fsdecode(TESTFN_UNDECODABLE)` and `TESTFN_UNENCODABLE` in some tests if you wish.

**Review by ruidazeng — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
UTF-8 can fail on surrogates. See #156610.

**Comment by serhiy-storchaka** [comment]
Opened #156713.

**Pull request merged**